powersimdata/input/converter/tests/test_pypsa_to_grid.py

```diff
@@ -36,19 +38,44 @@ def test_import_network_including_storages_from_pypsa_to_grid():
 @pytest.mark.skipif(find_spec("pypsa") is None, reason="Package PyPSA not available.")
 def test_import_exported_network():
 
-    ref = Grid("Western")
+    grid = Grid("Western")
+    ct = ChangeTable(grid)
+    storage = [
+        {"bus_id": 2021005, "capacity": 116.0},
+        {"bus_id": 2028827, "capacity": 82.5},
+        {"bus_id": 2028060, "capacity": 82.5},
+    ]
+    ct.add_storage_capacity(storage)
+    ref = TransformGrid(grid, ct.ct).get_grid()
+
     kwargs = dict(add_substations=True, add_load_shedding=False, add_all_columns=True)
     n = export_to_pypsa(ref, **kwargs)
-    test = FromPyPSA(n)
+    test = FromPyPSA(n, add_pypsa_cols=False)
 
     # Only a scaled version of linear cost term is exported to pypsa
     # Test whether the exported marginal cost is in the same order of magnitude
     ref_total_c1 = ref.gencost["before"]["c1"].sum()
     test_total_c1 = test.gencost["before"]["c1"].sum()
     assert ref_total_c1 / test_total_c1 > 0.95 and ref_total_c1 / test_total_c1 < 1.05
 
+    # Now overwrite costs
+    for c in ["c0", "c1", "c2"]:
+        test.gencost["before"][c] = ref.gencost["before"][c]
+        test.gencost["after"][c] = ref.gencost["after"][c]
+
     # Due to rounding errors we have to compare some columns in advance
     rtol = 1e-15
     assert_series_equal(ref.branch.x, test.branch.x, rtol=rtol)
     assert_series_equal(ref.branch.r, test.branch.r, rtol=rtol)
     assert_series_equal(ref.bus.Va, test.bus.Va, rtol=rtol)
+
+    test.branch.x = ref.branch.x
+    test.branch.r = ref.branch.r
+    test.bus.Va = ref.bus.Va
+
+    # storage specification is need in import but has to removed for testing
+    test.storage["gencost"].drop(columns="pypsa_component", inplace=True)
+    test.storage["gen"].drop(columns="pypsa_component", inplace=True)
+    test.storage["StorageData"].drop(columns="pypsa_component", inplace=True)
+
+    assert ref == test
```

Comment:
Not sure whether we should test over an imported Europe grid as well. Maybe we are not there yet given "make distinction to pypsa stores" is on the todo list.

Reply:
The roundtrip conversion `Network` --> `Grid` --> `Network` does not work as shown below:
```
>>> from powersimdata.network.europe_tub.model import TUB
>>> tub = TUB("Europe", reduction=1024)
Title: PyPSA-Eur: An Open Optimisation Model of the European Transmission System (Dataset)
Keywords: power system model, capacity expansion model, energy system model, electricity system model, python
Publication date: 2022-09-10
DOI: 10.5281/zenodo.6913032
Total size: 2075.6 MB

Link: https://zenodo.org/api/files/7528030f-2993-4f54-83e4-9fdd223b9ba5/networks.zip   size: 2075.6 MB
networks.zip is already downloaded correctly.
All files have been downloaded.
>>> tub.build()
INFO:pypsa.io:Imported network elec_s_1024_ec.nc has buses, carriers, generators, lines, links, loads, storage_units, stores
>>> from powersimdata.input.converter.pypsa_to_grid import FromPyPSA
>>> g = FromPyPSA(tub.network)
/Users/brdo/CEM/PowerSimData/powersimdata/input/converter/pypsa_to_grid.py:240: UserWarning: Substations could not be parsed.
  warnings.warn("Substations could not be parsed.")
>>> from powersimdata.input.exporter.export_to_pypsa import export_to_pypsa
>>> n = export_to_pypsa(g)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/brdo/CEM/PowerSimData/powersimdata/input/exporter/export_to_pypsa.py", line 128, in export_to_pypsa
    constants = grid.model_immutables.plants
```
The issue is that the `model_immutables` attribute of the created `Grid` object using `FromPyPSA` is never set:
```
>>> type(g.model_immutables)
```
It exists since `FromPyPSA` inherits from `AbstractGrid` but it is never set. However the `TUB` object has a `model_immutables` attribute:
```
>>> tub.model_immutables.plants
{'type2color': {'onwind': '#235ebc', 'offwind-ac': '#6895dd', 'offwind-dc': '#74c6f2', 'hydro': '#08ad97', 'PHS': '#08ad97', 'ror': '#4adbc8', 'solar': '#f9d002', 'biomass': '#0c6013', 'geothermal': '#ba91b1', 'OCGT': '#d35050', 'CCGT': '#b20101', 'nuclear': '#ff9000', 'coal': '#707070', 'lignite': '#9e5a01', 'oil': '#262626', 'H2': '#ea048a', 'battery': '#b8ea04'}, 'curtailable2color': {'solar': '#f9d002', 'onwind': '#235ebc', 'offwind-ac': '#6895dd', 'offwind-dc': '74c6f2'}, 'curtailable2hatchcolor': {'solar': '#808080', 'onwind': '#808080', 'offwind-ac': '#808080', 'offwind-dc': '#808080'}, 'efficiency': {'biomass': 0.416, 'OCGT': 0.39, 'CCGT': 0.55, 'coal': 0.33, 'lignite': 0.33, 'oil': 0.35}, 'carbon_per_mwh': {'biomass': 721, 'OCGT': 516, 'CCGT': 366, 'coal': 1021, 'lignite': 1212, 'oil': 762}, 'carbon_per_mbtu': {'biomass': 21.74, 'OCGT': 14.59, 'CCGT': 14.59, 'coal': 24.41, 'lignite': 28.97, 'oil': 19.32}, 'nox_per_mwh': {'biomass': 0.216, 'OCGT': 0.443, 'CCGT': 0.131, 'coal': 0.415, 'lignite': 0.415, 'oil': 9.689}, 'so2_per_mwh': {'biomass': 0.032, 'OCGT': 0.004, 'CCGT': 0.003, 'coal': 0.057, 'lignite': 0.115, 'oil': 0.237}, 'type2label': {'onwind': 'Onshore Wind', 'offwind-ac': 'Offshore Wind (AC)', 'offwind-dc': 'Offshore Wind (DC)', 'hydro': 'Reservoir & Dam', 'PHS': 'Pumped Hydro Storage', 'ror': 'Run of River', 'solar': 'Solar', 'biomass': 'Biomass', 'geothermal': 'Geothermal', 'OCGT': 'Open-Cycle Gas', 'CCGT': 'Combined-Cycle Gas', 'nuclear': 'Nuclear', 'coal': 'Coal', 'lignite': 'Lignite', 'oil': 'Oil', 'H2': 'Hydrogen Storage', 'battery': 'Battery Storage'}, 'curtailable2label': {'solar': 'Solar Curtailment', 'onwind': 'Onshore Wind Curtailment', 'offwind-ac': 'Offshore Wind Curtailment (AC)', 'offwind-dc': 'Offshore Wind Curtailment (DC)'}, 'label2type': {'Onshore Wind': 'onwind', 'Offshore Wind (AC)': 'offwind-ac', 'Offshore Wind (DC)': 'offwind-dc', 'Reservoir & Dam': 'hydro', 'Pumped Hydro Storage': 'PHS', 'Run of River': 'ror', 'Solar': 'solar', 'Biomass': 'biomass', 'Geothermal': 'geothermal', 'Open-Cycle Gas': 'OCGT', 'Combined-Cycle Gas': 'CCGT', 'Nuclear': 'nuclear', 'Coal': 'coal', 'Lignite': 'lignite', 'Oil': 'oil', 'Hydrogen Storage': 'H2', 'Battery Storage': 'battery'}, 'label2curtailable': {'Solar Curtailment': 'solar', 'Onshore Wind Curtailment': 'onwind', 'Offshore Wind Curtailment (AC)': 'offwind-ac', 'Offshore Wind Curtailment (DC)': 'offwind-dc'}, 'profile_resources': {'PHS', 'ror', 'offwind-dc', 'onwind', 'offwind-ac', 'solar', 'hydro'}, 'group_profile_resources': {'hydro': {'PHS', 'ror', 'hydro'}, 'solar': {'solar'}, 'wind': {'offwind-dc', 'onwind', 'offwind-ac'}}, 'curtailable_resources': {'PHS', 'ror', 'offwind-dc', 'onwind', 'offwind-ac', 'solar', 'hydro'}, 'thermal_resources': {'coal', 'OCGT', 'geothermal', 'nuclear', 'oil', 'biomass', 'lignite', 'CCGT'}, 'renewable_resources': {'offwind-dc', 'onwind', 'offwind-ac', 'solar'}, 'carbon_resources': {'biomass', 'coal', 'lignite', 'OCGT', 'CCGT', 'oil'}, 'clean_resources': {'onwind', 'geothermal', 'nuclear', 'hydro', 'ror', 'offwind-dc', 'offwind-ac', 'solar'}, 'group_all_resources': {'biomass': {'biomass'}, 'coal': {'coal', 'lignite'}, 'geothermal': {'geothermal'}, 'hydro': {'PHS', 'ror', 'hydro'}, 'ng': {'CCGT', 'OCGT'}, 'nuclear': {'nuclear'}, 'oil': {'oil'}, 'solar': {'solar'}, 'wind': {'offwind-dc', 'onwind', 'offwind-ac'}}, 'all_resources': {'coal', 'OCGT', 'onwind', 'geothermal', 'nuclear', 'oil', 'hydro', 'biomass', 'PHS', 'ror', 'lignite', 'offwind-dc', 'offwind-ac', 'CCGT', 'solar'}, 'pmin_as_share_of_pmax': {'biomass': None, 'CCGT': None, 'coal': None, 'geothermal': None, 'hydro': 0.9, 'lignite': None, 'nuclear': None, 'OCGT': None, 'oil': None, 'offwind-ac': 0, 'offwind-dc': 0, 'onwind': 0, 'PHS': 0.9, 'ror': None, 'solar': 0}}
```
We will need either to pass `model_immutables` to the constructor of `FromPyPSA` or add it to the created `Grid` object manually.

Reply:
@FabianHofmann, should we have a `if`/`else` logic on `model_immutables` in the `export_to_pypsa`?

Reply:
@rouille I think that is a good idea, the constants that are retrieved from `model_immmutables` are added to`Network.carriers` and not a necessary information. So I'd say that only if `model_immutables` is given, the values should be extracted.

Reply:
@FabianHofmann, can you take a look at the last [commit](ff254e79eb5606acc4407b31b2e8d4a0cc45f54c). It fixes the issue raised above. That said, the initial `Network` object and the one obtained after the roundtrip conversion are not equal:
```
>>> from powersimdata.network.europe_tub.model import TUB
>>> tub = TUB("Europe", reduction=1024)
Title: PyPSA-Eur: An Open Optimisation Model of the European Transmission System (Dataset)
Keywords: power system model, capacity expansion model, energy system model, electricity system model, python
Publication date: 2022-09-10
DOI: 10.5281/zenodo.6913032
Total size: 2075.6 MB

Link: https://zenodo.org/api/files/7528030f-2993-4f54-83e4-9fdd223b9ba5/networks.zip   size: 2075.6 MB
networks.zip is already downloaded correctly.
All files have been downloaded.
>>> tub.build()
INFO:pypsa.io:Imported network elec_s_1024_ec.nc has buses, carriers, generators, lines, links, loads, storage_units, stores
>>> from powersimdata.input.converter.pypsa_to_grid import FromPyPSA
>>> g = FromPyPSA(tub.network)
/Users/brdo/CEM/PowerSimData/powersimdata/input/converter/pypsa_to_grid.py:240: UserWarning: Substations could not be parsed.
  warnings.warn("Substations could not be parsed.")
>>> from powersimdata.input.exporter.export_to_pypsa import export_to_pypsa
>>> n = export_to_pypsa(g)
>>> len(n.generators)
6826
>>> len(tub.network.generators)
3754
>>> n.generators.columns
Index(['bus', 'Qmax', 'Qmin', 'Vg', 'mBase', 'p_nom', 'p_min_pu', 'ramp_q',
       'apf', 'carrier', 'interconnect', 'GenFuelCost', 'zone_id', 'zone_name',
       'lat', 'lon', 'pypsa_p_nom_max', 'pypsa_weight', 'pypsa_p_nom',
       'pypsa_capital_cost', 'pypsa_efficiency', 'pypsa_p_nom_min',
       'pypsa_p_nom_extendable', 'pypsa_p_max_pu', 'pypsa_bus',
       'pypsa_carrier', 'pypsa_control', 'pypsa_p_nom_opt', 'pypsa_type',
       'pypsa_p_min_pu', 'pypsa_p_set', 'pypsa_q_set', 'pypsa_sign',
       'pypsa_build_year', 'pypsa_lifetime', 'pypsa_committable',
       'pypsa_min_up_time', 'pypsa_min_down_time', 'pypsa_up_time_before',
       'pypsa_down_time_before', 'pypsa_ramp_limit_up',
       'pypsa_ramp_limit_down', 'pypsa_ramp_limit_start_up',
       'pypsa_ramp_limit_shut_down', 'committable', 'ramp_limit_down',
       'ramp_limit_up', 'start_up_cost', 'shut_down_cost', 'marginal_cost',
       'control', 'type', 'p_nom_extendable', 'p_nom_min', 'p_nom_max',
       'p_max_pu', 'p_set', 'q_set', 'sign', 'build_year', 'lifetime',
       'capital_cost', 'efficiency', 'min_up_time', 'min_down_time',
       'up_time_before', 'down_time_before', 'ramp_limit_start_up',
       'ramp_limit_shut_down', 'p_nom_opt'],
      dtype='object')
>>> tub.network.generators.columns
Index(['p_nom_max', 'weight', 'p_nom', 'capital_cost', 'efficiency',
       'p_nom_min', 'marginal_cost', 'p_nom_extendable', 'p_max_pu', 'bus',
       'carrier', 'control', 'p_nom_opt', 'type', 'p_min_pu', 'p_set', 'q_set',
       'sign', 'build_year', 'lifetime', 'committable', 'start_up_cost',
       'shut_down_cost', 'min_up_time', 'min_down_time', 'up_time_before',
       'down_time_before', 'ramp_limit_up', 'ramp_limit_down',
       'ramp_limit_start_up', 'ramp_limit_shut_down'],
      dtype='object')
``` 
What do you think? I would say that at least we should have the same number of generators in the two objects.

Reply:
This is fixed in #689.